### PR TITLE
weighter.py added

### PR DESCRIPTION
**Pull request by FilatovArtm**

### CODE REVIEW

File: weighter.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
+import numpy as np
+import os
+import ROOT as r
+import shlex, subprocess
+import argparse
+
+EOS_PATH = "root://eoslhcb.cern.ch/eos/ship/data/IS_experiment"
```
Comment (olantwin) [inline weighter.py]: `eoslhcb.cern.ch` is not used anymore by SHiP. We have moved to `eospublic.cern.ch//eos/experiment/ship/...`

File: weighter.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+FULL_MUONS_PATH = 'root://eoslhcb.cern.ch/eos/ship/data/Mbias/pythia8_Geant4-withCharm_onlyMuons_4magTarget.root'
```
Comment (olantwin) [inline weighter.py]: same here

File: weighter.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+SLAVE_CMD = '''python2 code/slave.py ''' \
+            '''--params {params} ''' \
+            '''-f /shield/worker_files/sampling_is/''' \
+            '''muons.root ''' \
+            '''--results /output/result.json ''' \
+            '''--hists /output/hists_{tag}.root --seed {seed} ''' \
+            '''--xs_path {xs_path}'''
+
+def loss(x):
```
Comment (olantwin) [inline weighter.py]: How do you handle the two muon polarities here?

File: weighter.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
+import numpy as np
+import os
+import ROOT as r
+import shlex, subprocess
+import argparse
+
+EOS_PATH = "root://eoslhcb.cern.ch/eos/ship/data/IS_experiment"
+FULL_MUONS_PATH = 'root://eoslhcb.cern.ch/eos/ship/data/Mbias/pythia8_Geant4-withCharm_onlyMuons_4magTarget.root'
+SLAVE_CMD = '''python2 code/slave.py ''' \
```
Comment (olantwin) [inline weighter.py]: The `\` are not needed. `'''` are combined automatically over several lines'''

File: weighter.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+    parser.add_argument('--results', default='results.json')
+    parser.add_argument('--hists', default='hists.root')
+    parser.add_argument('--params', required=True)
+    parser.add_argument('--point_id', required=True)
+    parser.add_argument('--tag', default="")
+    parser.add_argument('--seed', type=int, default=1)
+    parser.add_argument('--share_muons', type=float)
+
+    args = parser.parse_args()
+    args.xs_path = os.path.join(EOS_PATH, get_xs_path(args.tag, args.point_id))
```
Comment (olantwin) [inline weighter.py]: the slave can't write to EOS directly. EOS does not work like a normal file system (but can be mounted via fuse).

File: weighter.py [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+    for muon in intuple:
+        if i == indexes[indexes_pointer]:
+            outtuple.Fill(muon)
+            indexes_pointer += 1
+        if indexes_pointer == len(indexes):
+            break
+
+        i += 1
+
+    outtuple.Write()
```
Comment (olantwin) [inline weighter.py]: You should probably also close f.

File: weighter.py [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+def count_muons(filename_read):
+    '''
+    This function calculates number of muons in the file.
+    '''
+    f = r.TFile.Open(filename_read, 'read')
+    intuple = f.Get('pythia8-Geant4')
+    i = 0
+
+    for muon in intuple:
+        i += 1
```
Comment (olantwin) [inline weighter.py]: You could just use TNtuple.GetEntries() or TNtuple.GetEntriesFast() instead. Should be much faster as well.

File: weighter.py [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+
+def get_command_line(SLAVE_CMD, args):
+    return SLAVE_CMD.format(**vars(args))
+
+def main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        '-f',
+        '--input',
+        default='root://eoslhcb.cern.ch/'
```
Comment (olantwin) [inline weighter.py]: Again, EOS locations have changed. Forgot to adjust the other scripts, since i haven't used them since the change end of january.

**Review by olantwin — CHANGES_REQUESTED** [review]
The main problem is that you treat EOS like a normal filesystem (which it isn't). None of the normal os.path methods will work.

The other issues are pretty trivial to fix.

File: weighter.py (hunk 3 above)
Comment (FilatovArtm) [reply weighter.py]: Should I save muon polarities somewhere? or I can just look at the sign of x?

Comment (olantwin) [reply weighter.py]: The problem is that the formula for the loss depends on the muon polarity. Once it's calculated it is not needed anymore, but you need to be careful that it is taken into account for the calculation.

Comment (FilatovArtm) [reply weighter.py]: Yes, I understand, but how I can extract the information about polarities?

File: config.py [hunk 9]
```diff
@@ -109,16 +109,17 @@ def StripFreeParams(point):
         'near_run3',
         'cmd':
         '''/bin/bash -l -c 'source /opt/FairShipRun/config.sh; '''
-        '''python2 /code/slave.py '''
+        '''python2 /code/weighter.py '''
```
Comment (olantwin) [inline config.py]: I'd appreciate if you didn't change the general template. You could either modify it in the submission script or add a second template. I think I'd go with the first option.

Otherwise all the optimisation scripts will break.

File: weighter.py (hunk 3 above)
Comment (olantwin) [reply weighter.py]: Never mind, if you save the the `x` in the `xs` array, it has already been accounted for. But in general, via the PID: Muons are 13 and anti-muons are -13.

**Review by olantwin — CHANGES_REQUESTED** [review]
Looks mostly good to me. Apart from the template it's probably worth testing at this point.

File: config.py (hunk 9 above)
Comment (FilatovArtm) [reply config.py]: I just thought that we could use separate branches. But okay, I'll create another template.

Comment (olantwin) [reply config.py]: I think maintaining two separate branches for this would get hard to maintain. I especially would really like to avoid having more than one good docker image at a time.

But instead of creating a separate template, it's probably easiest if you just modify the necessary fields in the script that submits your docker jobs, e.g.
```python3
template['cmd']='''/bin/bash -l -c 'source /opt/FairShipRun/config.sh; '''
'''python2 /code/weighter.py ... '''
```

Comment (FilatovArtm) [reply config.py]: Okay, I agree!

**Comment by olantwin** [comment]
BTW, do pylint/flake8 run without warnings on your code?

File: importance_sampling_optimizer/config.py [hunk 10]
```diff
@@ -0,0 +1,41 @@
+
+IMAGE = 'olantwin/ship-shield'
+IMAGE_TAG =
+SLEEP_TIME = 60
+
+JOB_TEMPLATE_IMP_SAMPLING = {
+    'input': ['eos:/eos/experiment/ship/skygrid/importance_sampling'],
```
Comment (olantwin) [inline importance_sampling_optimizer/config.py]: Does this mount or copy?

File: importance_sampling_optimizer/config.py [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+        'max_memoryMB':
+        1024,
+        'min_memoryMB':
+        512,
+        'run_id':
+        'near_run3',
+        'cmd':
+        '''python2 /code/weighter.py '''
+        '''--params {params} '''
+        '''-f /shield/worker_files/sampling_1/''' #TODO: write here the path to the file with all muons
```
Comment (olantwin) [inline importance_sampling_optimizer/config.py]: `root://eospublic.cern.ch//eos/experiment/ship/data/Mbias/pythia8_Geant4-withCharm_onlyMuons_4magTarget.root`

File: result_collector/aggregate_results.py [hunk 12]
```diff
@@ -0,0 +1,59 @@
+import numpy as np
+import re
+
+def get_number(filename):
+    return re.findall(r'\d+', filename)[0]
```
Comment (olantwin) [inline result_collector/aggregate_results.py]: Could this be done without a regular expression? What does filename look like?

File: result_collector/aggregate_results.py [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+
+def load_previous_cumulative_arrays():
```
Comment (olantwin) [inline result_collector/aggregate_results.py]: Again, I'd suggest looking into ROOT IO, especially into TChain as you might deal with many small files on EOS.

File: result_collector/aggregate_results.py [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+    cum_loss = np.load("/input/cumloss.npy")
+    cum_indeces = np.load("/input/cumindeces.npy")
+    return cum_loss, cum_indeces
+
+def load_previous_results(tag):
```
Comment (olantwin) [inline result_collector/aggregate_results.py]: Do you delete the files that are already processed?

File: weighter.py [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+
+def get_command_line(SLAVE_CMD, args):
+    return SLAVE_CMD.format(**vars(args))
+
+def main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        '-f',
+        '--input',
+        default='/shield/worker_files/sampling_1/muons_1.root')
```
Comment (olantwin) [inline weighter.py]: This path needs adjusting, I think...

**Review by olantwin — CHANGES_REQUESTED** [review]
Much better!

The main question I still have is: How do all the files/scripts play together, i.e. which script controls everything, what runs in docker, and in which order/structure?

File: result_collector/aggregate_results.py (hunk 14 above)
Comment (FilatovArtm) [reply result_collector/aggregate_results.py]: It is good idea. But unfortunately docker doesn't have access to eos directly

File: result_collector/aggregate_results.py (hunk 13 above)
Comment (FilatovArtm) [reply result_collector/aggregate_results.py]: I would suggest to leave it now, and in case of stable work, we can rewrite that part.

File: result_collector/aggregate_results.py (hunk 12 above)
Comment (FilatovArtm) [reply result_collector/aggregate_results.py]: file names look like xs_tag<number>.npy

Comment (FilatovArtm) [reply result_collector/aggregate_results.py]: Yes, probably we can, I just did it in the most straightforward way. Probably we can add one more '_' and split by that

File: result_collector/aggregate_results.py (hunk 14 above)
Comment (olantwin) [reply result_collector/aggregate_results.py]: It might then make sense to have a script we run manually on lxplus to clean up every now and again.

Also, how do you keep track of what has been processed?

**Comment by FilatovArtm** [comment]
The general structure is the following:
important_sampling_optimizer.py starts the docker with the weighter.py
weighter.py starts slave.py
// batch is calculated
important_sampling_optimizer.py starts result_collector

File: importance_sampling_optimizer/config.py [hunk 16]
```diff
@@ -4,7 +4,8 @@
 SLEEP_TIME = 60
 
 JOB_TEMPLATE_IMP_SAMPLING = {
-    'input': ['eos:/eos/experiment/ship/skygrid/importance_sampling'],
+    'input': ['eos:/eos/experiment/ship/skygrid/importance_sampling',
+              'eos:/eos/ship/data/Mbias'],
```
Comment (olantwin) [inline importance_sampling_optimizer/config.py]: 1. We don't want to copy all the files in there! Most of the are pretty large and not needed.
2. It's `/eos/experiment/ship`, not `/eos/ship`
3. @sashabaranov : Does the `eos` loader use `eospublic.cern.ch` or still `eoslhcb.cern.ch`?

File: result_collector/aggregate_results.py (hunk 13 above)
Comment (olantwin) [reply result_collector/aggregate_results.py]: Good point. For a first version this should be fine.

**Comment by FilatovArtm** [comment]
Now, I am fixing the issues with templates.

File: result_collector/aggregate_results.py (hunk 12 above)
Comment (olantwin) [reply result_collector/aggregate_results.py]: How about `os.path.splitext(os.path.basename(filename).split('_')[0])[0]`? or something like that?

Edit: Sorry, didn't see your other comment concerning this.

File: importance_sampling_optimizer/config.py (hunk 16 above)
Comment (sashabaranov) [reply importance_sampling_optimizer/config.py]: @olantwin 

1. Yes, we dont want this.
2. For production output we use `eos:/eos/experiment/ship/skygrid/background-prod-2018/$JOB_ID`, so maybe it is right
3. it uses ` root://eospublic.cern.ch/ `

Comment (olantwin) [reply importance_sampling_optimizer/config.py]: Thanks for the quick reply!
1. So maybe we should copy the file to the cluster once and then access it locally? Any other suggestions?
2. `/eos/experiment` is right. This was concerning `eos:/eos/ship/data/Mbias` which won't work
3. Perfect :)

Comment (sashabaranov) [reply importance_sampling_optimizer/config.py]: > So maybe we should copy the file to the cluster once and then access it locally? Any other suggestions?

If you'll need to copy something — just tell me what.

Comment (olantwin) [reply importance_sampling_optimizer/config.py]: I'd say we need `root://eospublic.cern.ch//eos/experiment/ship/data/Mbias/pythia8_Geant4-withCharm_onlyMuons_4magTarget.root` somewhere where it can be easily be accessed by the docker workers. Is that possible? Maybe where we'd normally keep the input files for the workers?

**Comment by olantwin** [comment]
Does every worker need to create its own muon sample from the full sample?

**Review by FilatovArtm — COMMENTED** [review]
@sashabaranov is the path correct now? will it just copy the file to input?

File: importance_sampling/importance_sampling_optimizer.py [hunk 17]
```diff
@@ -25,42 +27,40 @@ def main():
         type=int
     )
     args = parser.parse_args()
-    tag = f'important_sampling_{args.opt}' + f'_{args.tag}' if args.tag else ''
+    tag = "important_sampling_{args.opt}_{args.tag}".format(**args)
```
Comment (olantwin) [inline importance_sampling/importance_sampling_optimizer.py]: Why this change?

**Review by olantwin — APPROVED** [review]

File: slave.py [hunk 18]
```diff
@@ -195,6 +196,7 @@ def main():
             chain = r.TChain('cbmsim')
             chain.Add(outFile)
             xs = analyse(chain, args.hists)
+            np.save(args.xs_path, np.array(xs))
```
Comment (olantwin) [inline slave.py:199]: I think you're still missing the information which muons created non-zero entries in `xs`, no?

Comment (olantwin) [reply slave.py:199]: In [analysis.py](analysis.py) you need to add a second case and then push-back `w=0` into `xs` in the case where the muon does not get a non-zero weight.

**Review by olantwin — CHANGES_REQUESTED** [review]
Need to save weight for all muons, even if zero.

File: slave.py (hunk 18 above)
Comment (FilatovArtm) [reply slave.py:199]: what does xs contains? list of non - zero 'x' coordinates?

Comment (olantwin) [reply slave.py:199]: A list of weights of muons, but only for those with a non-zero weight. They are then summed to get the muon contribution to the loss function.

Comment (FilatovArtm) [reply slave.py:199]: okay, I'll add this feature to analysis.py

**Review by olantwin — APPROVED** [review]
